cdap-app-templates/cdap-program-report/cdap-program-report1_2.10/src/main/scala/co/cask/cdap/report/ReportAggregationFunction.scala

```diff
@@ -0,0 +1,248 @@
+  import ReportAggregationFunction._
+
+  override def inputSchema: StructType = new StructType()
+    .add(Constants.NAMESPACE, StringType, false)
+    .add(Constants.APPLICATION_NAME, StringType, false)
+    .add(Constants.APPLICATION_VERSION, StringType, false)
+    .add(Constants.PROGRAM_TYPE, StringType, false)
+    .add(Constants.PROGRAM, StringType, false)
+    .add(Constants.RUN, StringType, false)
+    .add(Constants.STATUS, StringType, false)
+    .add(Constants.TIME, LongType, false)
+    .add(Constants.MESSAGE_ID, StringType, false)
+    .add(Constants.START_INFO, INPUT_START_INFO_SCHEMA, true)
+
+  override def bufferSchema: StructType = new StructType()
+    .add(Constants.NAMESPACE, StringType, false)
+    .add(Constants.APPLICATION_NAME, StringType, false)
+    .add(Constants.APPLICATION_VERSION, StringType, false)
+    .add(Constants.PROGRAM_TYPE, StringType, false)
+    .add(Constants.PROGRAM, StringType, false)
+    .add(Constants.RUN, StringType, false)
+    .add(STATUSES, ArrayType(new StructType()
+      .add(Constants.STATUS, StringType, false)
+      .add(Constants.TIME, LongType)), false)
+    .add(Constants.START_INFO, BUFFER_START_INFO_SCHEMA,
+      true)
+
+  override def dataType: DataType = new StructType()
+    .add(Constants.NAMESPACE, StringType, false)
+    .add(Constants.ARTIFACT_NAME, StringType, true)
+    .add(Constants.ARTIFACT_SCOPE, StringType, true)
+    .add(Constants.ARTIFACT_VERSION, StringType, true)
+    .add(Constants.APPLICATION_NAME, StringType, false)
+    .add(Constants.APPLICATION_VERSION, StringType, false)
+    .add(Constants.PROGRAM_TYPE, StringType, false)
+    .add(Constants.PROGRAM, StringType, false)
+    .add(Constants.RUN, StringType, false)
+    .add(Constants.STATUS, StringType, false)
+    .add(Constants.START, LongType, true)
+    .add(Constants.RUNNING, LongType, true)
+    .add(Constants.END, LongType, true)
+    .add(Constants.DURATION, LongType, true)
+    .add(Constants.USER, StringType, true)
+    .add(Constants.START_METHOD, StringType, false)
+    .add(Constants.RUNTIME_ARGUMENTS, MapType(StringType, StringType), true)
+    .add(Constants.NUM_LOG_WARNINGS, IntegerType, true)
+    .add(Constants.NUM_LOG_ERRORS, IntegerType, true)
+    .add(Constants.NUM_RECORDS_OUT, IntegerType, true)
+
+  override def deterministic: Boolean = true
+
+  override def initialize(buffer: MutableAggregationBuffer): Unit = {
+    val bufferRow = new GenericRowWithSchema(buffer.toSeq.toArray, bufferSchema)
+    buffer.update(bufferRow.fieldIndex(Constants.NAMESPACE), "")
+    buffer.update(bufferRow.fieldIndex(Constants.APPLICATION_NAME), "")
+    buffer.update(bufferRow.fieldIndex(Constants.APPLICATION_VERSION), "")
+    buffer.update(bufferRow.fieldIndex(Constants.PROGRAM_TYPE), "")
+    buffer.update(bufferRow.fieldIndex(Constants.PROGRAM), "")
+    buffer.update(bufferRow.fieldIndex(Constants.RUN), "")
+    buffer.update(bufferRow.fieldIndex(STATUSES), Seq.empty[Row])
```

Comment:
use a similar way to initialize this using a list. maybe create a list `singleValueFields` outside and use that in both `initialize` and `update` and then initialize status separately after that.